generator/cells_merger.cpp

```diff
@@ -0,0 +1,273 @@
+    }
+  }
+  // Top left
+  for (int32_t x = 0; x <= m_maxX; ++x)
+  {
+    for (int32_t y = m_maxY; y >= 0; --y)
+    {
+      if (!Has(x, y))
+        continue;
+      auto & cell = Get(x, y);
+      cell.SetTopLeft(Min(TryGet(x - 1, y).GetTopLeft(),
+                          TryGet(x, y + 1).GetTopLeft(),
+                          TryGet(x - 1, y + 1).GetTopLeft()) + 1);
+    }
+  }
+  // Top right
+  for (int32_t x = m_maxX; x >= 0 ; --x)
+  {
+    for (int32_t y = m_maxY; y >= 0; --y)
+    {
+      if (!Has(x, y))
+        continue;
+      auto & cell = Get(x, y);
+      cell.SetTopRight(Min(TryGet(x + 1, y).GetTopRight(),
+                           TryGet(x, y + 1).GetTopRight(),
+                           TryGet(x + 1, y + 1).GetTopRight()) + 1);
+    }
+  }
+  for (auto & pair : m_matrix)
+    pair.second.CalcSum();
+}
+
+CellWrapper & CellsMerger::Get(const m2::PointI & xy)
+{
+  auto it = m_matrix.find(xy);
+  CHECK(it != std::cend(m_matrix), ());
+  return it->second;
+}
+
+CellWrapper & CellsMerger::Get(int32_t x, int32_t y)
+{
+  return Get({x, y});
+}
+
+CellWrapper const & CellsMerger::TryGet(int32_t x, int32_t y, CellWrapper const & default_) const
+{
+  auto const it = m_matrix.find({x, y});
+  return it == std::cend(m_matrix) ? default_ : it->second;
+}
+
+bool CellsMerger::MayTransform(m2::PointI const & minXy, m2::PointI const & maxXy)
+{
+  auto const xMin = std::min(minXy.x, maxXy.x);
+  auto const xMax = std::max(minXy.x, maxXy.x);
+  auto const yMin = std::min(minXy.y, maxXy.y);
+  auto const yMax = std::max(minXy.y, maxXy.y);
+  for (int32_t x = xMin; x <= xMax; ++x)
+  {
+    for (int32_t y = yMin; y <= yMax; ++y)
+    {
+      if (!Has(x, y))
+        return false;
+    }
+  }
+  return true;
+}
+
+m2::PointI CellsMerger::FindDirection(m2::PointI const & startXy) const
+{
+  std::array<std::pair<size_t, m2::PointI>, 4> directions;
+  auto const topRight = TryGet(startXy.x + 1, startXy.y).GetSum() +
+                        TryGet(startXy.x, startXy.y + 1).GetSum() +
+                        TryGet(startXy.x + 1, startXy.y + 1).GetSum();
+  directions[0] = {topRight, m2::PointI{startXy.x + 1, startXy.y + 1}};
+  auto const topLeft = TryGet(startXy.x - 1, startXy.y).GetSum() +
+                       TryGet(startXy.x, startXy.y + 1).GetSum() +
+                       TryGet(startXy.x - 1, startXy.y + 1).GetSum();
+  directions[1] = {topLeft, m2::PointI{startXy.x - 1, startXy.y + 1}};
+  auto const bottomRight = TryGet(startXy.x + 1, startXy.y).GetSum() +
+                           TryGet(startXy.x, startXy.y - 1).GetSum() +
+                           TryGet(startXy.x + 1, startXy.y - 1).GetSum();
+  directions[2] = {bottomRight, m2::PointI{startXy.x + 1, startXy.y - 1}};
+  auto const bottomLeft = TryGet(startXy.x - 1, startXy.y).GetSum() +
+                          TryGet(startXy.x, startXy.y - 1).GetSum() +
+                          TryGet(startXy.x - 1, startXy.y - 1).GetSum();
```

Comment:
Can we lower the amount of hardcode here?

Reply:
++